File: DemoMMOUnity/Assets/Game/Net/Script/Net/CsProtocol/Protocol/Map/PuzzleChapter.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using ZJYFrameWork.Collection.Reference;
using ZJYFrameWork.Net.Core;

namespace ZJYFrameWork.Net.CsProtocol.Buffer.Protocol.Map
{
    /// <summary>
    /// 地图 章节数据
    /// </summary>
    public class PuzzleChapter : Model, IPacket, IReference
    {
        /**
         * id
         */
        public int id { get; set; }

        /**
         * 章节名
         */
        public string chapterName { get; set; }

        /**
         * 当前章节 最小 关卡id
         */
        public int minPuzzle { get; set; }

        /**
         * 当前章节 最大 关卡id
         */
        public int maxPuzzle { get; set; }

        /**
         * 当前 进行中的 关卡id
         * <p>
         * 如果这个数字 为负数就查看 doneMaxPuzzleId
         * 是否有值，没有值就代表 没有进行
         * </p>
         */
        public int nowCarryOutPuzzleId { get; set; }

        /**
         * 完成的最大关卡id 只有 关卡完成之后 才会更新
         */
        public int doneMaxPuzzleId { get; set; }

        public short ProtocolId()
        {
            return 204;
        }

        public void Clear()
        {
            id = -1;
            chapterName = string.Empty;
            minPuzzle = -1;
            maxPuzzle = -1;
            nowCarryOutPuzzleId = -1;
            doneMaxPuzzleId = -1;
        }

        public static PuzzleChapter ValueOf()
        {
            var data = ReferenceCache.Acquire<PuzzleChapter>();
            return data;
        }
    }

    public class PuzzleChapterRegistration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return 204;
        }

        public void Write(ByteBuffer buffer, IPacket packet)
        {
            if (packet == null)
            {
                return;
            }

            var data = (PuzzleChapter)packet;
            var packetData = new ServerMessageWrite(protocolId: ProtocolId(), data);
            var jsonString = JsonConvert.SerializeObject(packetData);
            buffer.WriteString(jsonString);
        }

        public IPacket Read(string json = "")
        {
            var packet = PuzzleChapter.ValueOf();
            packet.Clear();
            if (string.IsNullOrEmpty(json))
            {
                return packet;
            }

            //解析
            var dict = JsonConvert.DeserializeObject<Dictionary<object, object>>(json);
            foreach (var (key, value) in dict)
            {
                var keyStr = key.ToString();
                var valueStr = value.ToString();
                switch (keyStr)
                {
                    case "id":
                    {
                        var id = int.Parse(valueStr);
                        packet.id = id;
                    }
                        break;
                    case "chapterName":
                    {
                        packet.chapterName = valueStr;
                    }
                        break;
                    case "minPuzzle":
                    {
                        var minPuzzle = int.Parse(valueStr);
                        packet.minPuzzle = minPuzzle;
                    }
                        break;
                    case "maxPuzzle":
                    {
                        var maxPuzzle = int.Parse(valueStr);
                        packet.maxPuzzle = maxPuzzle;
                    }
                        break;
                    case "nowCarryOutPuzzleId":
                    {
                        var nowCarryOutPuzzleId = int.Parse(valueStr);
                        packet.nowCarryOutPuzzleId = nowCarryOutPuzzleId;
                    }
                        break;
                    case "doneMaxPuzzleId":
                    {
                        var doneMaxPuzzleId = int.Parse(valueStr);
                        packet.doneMaxPuzzleId = doneMaxPuzzleId;
                    }
                        break;
                }
            }

            return packet;
        }
    }
}

File: UnityMMOClient/Lua/application/app/net/webSocket/luaProtocol/puzzle/puzzleChapter.lua
---@class PuzzleChapter
local PuzzleChapter = class("PuzzleChapter")
function PuzzleChapter:ctor()
    --- 章节名
    ---@type string
    self.chapterName = string.empty
    --- 完成的最大关卡id 只有 关卡完成之后 才会更新
    ---@type number
    self.doneMaxPuzzleId = 0
    --- id
    ---@type number
    self.id = 0
    --- 当前章节 最大 关卡id
    ---@type number
    self.maxPuzzle = 0
    --- 当前章节 最小 关卡id
    ---@type number
    self.minPuzzle = 0
    --- 当前 进行中的 关卡id
    ---@type number
    self.nowCarryOutPuzzleId = 0
end

---@param chapterName string 章节名
---@param doneMaxPuzzleId number 完成的最大关卡id 只有 关卡完成之后 才会更新
---@param id number id
---@param maxPuzzle number 当前章节 最大 关卡id
---@param minPuzzle number 当前章节 最小 关卡id
---@param nowCarryOutPuzzleId number 当前 进行中的 关卡id
---@return PuzzleChapter
function PuzzleChapter:new(chapterName, doneMaxPuzzleId, id, maxPuzzle, minPuzzle, nowCarryOutPuzzleId)
    self.chapterName = chapterName --- java.lang.String
    self.doneMaxPuzzleId = doneMaxPuzzleId --- int
    self.id = id --- int
    self.maxPuzzle = maxPuzzle --- int
    self.minPuzzle = minPuzzle --- int
    self.nowCarryOutPuzzleId = nowCarryOutPuzzleId --- int
    return self
end

---@return number
function PuzzleChapter:protocolId()
    return 204
end

---@return string
function PuzzleChapter:write()
    local message = {
        protocolId = self:protocolId(),
        packet = {
            chapterName = self.chapterName,
            doneMaxPuzzleId = self.doneMaxPuzzleId,
            id = self.id,
            maxPuzzle = self.maxPuzzle,
            minPuzzle = self.minPuzzle,
            nowCarryOutPuzzleId = self.nowCarryOutPuzzleId
        }
    }
    local jsonStr = JSON.encode(message)
    return jsonStr
end

---@return PuzzleChapter
function PuzzleChapter:read(data)

    local packet = self:new(
            data.chapterName,
            data.doneMaxPuzzleId,
            data.id,
            data.maxPuzzle,
            data.minPuzzle,
            data.nowCarryOutPuzzleId)
    return packet
end

--- 章节名
---@type  string 章节名
function PuzzleChapter:getChapterName()
    return self.chapterName
end

--- 完成的最大关卡id 只有 关卡完成之后 才会更新
---@return number 完成的最大关卡id 只有 关卡完成之后 才会更新
function PuzzleChapter:getDoneMaxPuzzleId()
    return self.doneMaxPuzzleId
end
--- id
---@return number id
function PuzzleChapter:getId()
    return self.id
end
--- 当前章节 最大 关卡id
---@return number 当前章节 最大 关卡id
function PuzzleChapter:getMaxPuzzle()
    return self.maxPuzzle
end
--- 当前章节 最小 关卡id
---@return number 当前章节 最小 关卡id
function PuzzleChapter:getMinPuzzle()
    return self.minPuzzle
end
--- 当前 进行中的 关卡id
---@return number 当前 进行中的 关卡id
function PuzzleChapter:getNowCarryOutPuzzleId()
    return self.nowCarryOutPuzzleId
end


return PuzzleChapter
